### PR TITLE
sql/parser: Print datum dates/timestamps in apostrophes

### LINKED ISSUE #12420: sql: VIEWs don't put quotations around date constants
This causes selection on the view to fail.
For instance:

```
CREATE TABLE foo (k DATE NOT NULL);
CREATE VIEW bar AS SELECT k FROM foo WHERE k = date '1988-11-12';
SELECT * FROM bar;
pq: failed to parse underlying query from view "bar": syntax error at or near "1988"
SELECT k FROM foo WHERE k = DATE 1988-11-12
```

### PR DESCRIPTION
**Pull request by a-robinson**
Without this, they get printed out in a format that prevents them from
being properly parsed. This was only hit in a very particular edge case
of using a literal to compare against a DATE column in a view's select
statement.

Fixes #12420.

It's a little concerning that this appears to change the format of keys...

Also, I wonder whether we may want to create the persisted string representation of the query during the construction of the `createViewNode` rather than in its `Start` method to avoid other quirks of the transformations that may be applied to the AST between the two. What do you think, @knz?

Finally, I'm still tracking down the last test failure (`TestCopyRandom`, which is still failing due to how this affects placeholder queries), but sending out now for faster feedback on everything else.

@arjunravinarayan

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12450)

### CODE REVIEW

**Comment by a-robinson** [comment]
Ok, all tests now passing. `TestCopyRandom` was failing because lib/pq puts apostrophes around timestamp literals, so the bug was just in the test.

**Comment by knz** [comment]
Yes making a string rendering earlier seems to be a good idea.

LGTM otherwise

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12450#-:-KZ7ZVhS0d9ksRKWhzCz:b-ikv5g9)*

**Comment by dt** [comment]
fwiw, it's only changing the _pretty print_ of the keys as used by DebugVals -- which for a scan node is filled by RowFetcher's prettyEncDatums, which in turn is just implicitly String() ing Datum, thus getting your change to FormatNode.

**Comment by a-robinson** [comment]
Phew, thanks @dt!

I've added a commit to move the view source query formatting to the planNode's constructor.

**Comment by arjunravinarayan** [comment]
:lgtm: Thanks for getting to this!

---

Reviewed 8 of 8 files at r1.
Review status: 8 of 9 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12450#-:-KZ7kBw4CeSeVQTURNyq:b-s5c2iv)*

File: pkg/sql/parser/eval.go [hunk 1]
```diff
@@ -1930,8 +1930,13 @@ func (expr *CastExpr) Eval(ctx *EvalContext) (Datum, error) {
 	case *StringColType:
 		var s DString
 		switch t := d.(type) {
-		case *DBool, *DInt, *DFloat, *DDecimal, *DTimestamp, *DTimestampTZ, *DDate, dNull:
+		case *DBool, *DInt, *DFloat, *DDecimal, dNull:
 			s = DString(d.String())
+		case *DTimestamp, *DTimestampTZ, *DDate:
+			var buf bytes.Buffer
+			d.Format(&buf, FmtBareStrings)
+			fmt.Println(buf.String())
```
Comment (tamird) [inline pkg/sql/parser/eval.go]: what's this?

Comment (a-robinson) [inline pkg/sql/parser/eval.go]: Ugggghhhh thanks

File: pkg/sql/copy_in_test.go [hunk 2]
```diff
@@ -178,7 +179,9 @@ func TestCopyRandom(t *testing.T) {
 		row[1] = time.Duration(rng.Int63()).String()
 		for j, t := range types {
 			d := sqlbase.RandDatum(rng, sqlbase.ColumnType{Kind: t}, false)
-			ds := d.String()
+			var buf bytes.Buffer
+			d.Format(&buf, parser.FmtBareStrings)
+			ds := buf.String()
```
Comment (a-robinson) [inline pkg/sql/copy_in_test.go]: Thanks, added that onto #12454

**Pull request merged**